Feature/analyzer documentation (#1384)
microsoft/presidio: docs/analyzer/recognizer_registry_provider.md

# Customizing recognizer registry
To load recognizers from file, use `RecognizerRegistryProvider` to instantiate the recognizer registry and then pass it through to the analyzer engine:

```python
from presidio_analyzer import AnalyzerEngine
from presidio_analyzer.recognizer_registry import RecognizerRegistryProvider

recognizer_registry_conf_file = "./analyzer/recognizers-config.yml"

provider = RecognizerRegistryProvider(
                conf_file=recognizer_registry_conf_file
            )
registry = provider.create_recognizer_registry()
analyzer = AnalyzerEngine(
            registry=registry,
            )

results = analyzer.analyze(text="My name is Morris")
print(results)
```

## Configuration file structure

```yaml
global_regex_flags: 26
recognizers: 
...
```

The configuration file consists of two parts:

  - `global_regex_flags`: regex flags to be used in regex matching.
  - `recognizers`: a list of recognizers to be loaded by the recognizer registry. This list consists of two different types of recognizers: 

>     - Predefined: A set of already defined recognizer classes in presidio

Reviewer: Predefined could also include recognizer classes defined in code by the user (that inherit from `EntityRecognizer`)

>     - Custom: custom created pattern recognizers that are created ad hoc based on the fields provided in the configuration file

Reviewer: Ad-hoc recognizer is a different thing in Presidio- a recognizer you can add to the request without having to register it in the registry. I would suggest using a different terminology here to avoid confusion.

Author: Better to just discard it then

## Recognizer list

The recognizer list comprises of both the predefined and custom recognizers, for example: 

```yaml
...
    - name: CreditCardRecognizer
    supported_languages:
    - language: en
        context: [credit, card, visa, mastercard, cc,
        amex, discover, jcb, diners, maestro, instapayment]
    - language: es
        context: [tarjeta, credito, visa, mastercard, cc,
        amex, discover, jcb, diners, maestro, instapayment]
    - language: it
    - language: pl
    type: predefined

    - name: UsBankRecognizer
    supported_languages: 
    - en
    type: predefined

    - MedicalLicenseRecognizer

    - name: ExampleCustomRecognizer
    patterns:
        - name: "zip code (weak)"
        regex: "(\\b\\d{5}(?:\\-\\d{4})?\\b)"
        score: 0.01
        - name: "zip code (weak)"
        regex: "(\\b\\d{5}(?:\\-\\d{4})?\\b)"
        score: 0.01
    supported_languages:
        - language: en

>         context: [credit]

Reviewer: can we use a different context word here? `credit` could be confusing as we're detecting a zip code here.

Reviewer: In the english context too

        - language: es
        context: [tarjeta,credito]
    supported_entity: "ZIP"

>     deny_list: [Mr., Mrs., Ms., Miss, Dr., Prof.]

Reviewer: I would remove the deny_list part, or put it in a third recognizer. It's confusing to have both the zip regex and the titles deny_list.

Author: I didn't want to create a large yml, but this recognizer is too bloated :D

    deny_list_score: 1
    type: custom
    enabled: true
```

The recognizer parameters:

>   - `supported_languages`: A list of supported languages that the analyzer will support. In case this field is missing, a recognizer will be created for each supported language provided be `AnalyzerEngine`. 

Reviewer: ```suggestion
  - `supported_languages`: A list of supported languages that the analyzer will support. In case this field is missing, a recognizer will be created for each supported language provided to the `AnalyzerEngine`. 
```

>   In addition to the language code, this field also contains the context, which increases confidence in the detection (as seen in the credit card example above).

Reviewer: ```suggestion
  In addition to the language code, this field also contains a list of context words, which increases confidence in the detection in case it is found in the surroundings of a detected entity (as seen in the credit card example above).
```

  - `type`: this could be either predefined or custom.
  - `name`: Different per the type of the recognizer. For predefined recognizers, this is the class name as defined in presidio, while for custom recognizers, it will be set as the name of the recognizer. Note: for predefined recognizers, it's possible to provide the class name to instantiate the recognizer - see `MedicalLicenseRecognizer` above.

>   - `patterns`: Pattern array that contains a name, score and regex that define matching patterns.

Reviewer: ```suggestion
  - `patterns`: a list of objects of type `Pattern` that contains a name, score and regex that define matching patterns.
```

provider = RecognizerRegistryProvider(
                conf_file=recognizer_registry_conf_file
            )
registry = provider.create_recognizer_registry()
analyzer = AnalyzerEngine(
            registry=registry,
            )

results = analyzer.analyze(text="My name is Morris", language="en")
print(results)
```

## Configuration file structure

```yaml
global_regex_flags: 26
recognizers: 
...
```

The configuration file consists of two parts:

  - `global_regex_flags`: regex flags to be used in regex matching.
  - `recognizers`: a list of recognizers to be loaded by the recognizer registry. This list consists of two different types of recognizers: 
    - Predefined: A set of already defined recognizer classes in presidio. This includes all recognizers defined in the codebase (along with user defined recognizers) that inherit from EntityRecognizer.
    - Custom: custom created pattern recognizers that are created based on the fields provided in the configuration file.

## Recognizer list

The recognizer list comprises of both the predefined and custom recognizers, for example: 

```yaml
...
    - name: CreditCardRecognizer

>     supported_languages:

Reviewer: same here. yaml formatting seems off

    - language: en
        context: [credit, card, visa, mastercard, cc,

>         amex, discover, jcb, diners, maestro, instapayment]

Reviewer: list has to be on the same line to be valid

    - language: es
        context: [tarjeta, credito, visa, mastercard, cc,
        amex, discover, jcb, diners, maestro, instapayment]
    - language: it
    - language: pl
    type: predefined

    - name: UsBankRecognizer
    supported_languages: 
    - en
    type: predefined

    - MedicalLicenseRecognizer

    - name: ExampleCustomRecognizer
    patterns:
        - name: "zip code (weak)"
        regex: "(\\b\\d{5}(?:\\-\\d{4})?\\b)"
        score: 0.01
        - name: "zip code (weak)"
        regex: "(\\b\\d{5}(?:\\-\\d{4})?\\b)"
        score: 0.01
    supported_languages:
        - language: en
        context: [credit]
        - language: es

>         context: [zip, code]

Reviewer: [código, postal]

        - language: en
        context: [credit]
        - language: es
        context: [zip, code]
    supported_entity: "ZIP"
    type: custom
    enabled: true

    - name: "TitlesRecognizer"
    supported_language: "en"
    supported_entity: "TITLE"
    deny_list: [Mr., Mrs., Ms., Miss, Dr., Prof.]
    deny_list_score: 1
```

The recognizer parameters:

  - `supported_languages`: A list of supported languages that the analyzer will support. In case this field is missing, a recognizer will be created for each supported language provided to the `AnalyzerEngine`. 
  In addition to the language code, this field also contains a list of context words, which increases confidence in the detection in case it is found in the surroundings of a detected entity (as seen in the credit card example above).

>   - `type`: this could be either predefined or custom. As this is optional, if not stated otherwise, the default type is custom.

Reviewer: Is it custom? If I write a name of a recognizer (say, CreditCardRecognizer), wouldn't it load the predefined one?

Author: You're right! It goes like this: 
1. A single string is predefined because it'll try and load the class instance with defaults
2. once you add up parameters the recognizer is custom by default. The reason for it is backward compatibility - previously everything was custome, and we don't wanna break the functionality.

Reviewer: Actually, this is a little confusing. Let's think of a way to elaborate the difference. I already see users getting confused around this flag.
